File: be-nodejs/src/libs/response.ts
import { Response, Request } from 'express';
import { config } from '@/configs/config';
import { t, changeLanguage } from '@/configs/i18n';

export interface ApiResponse<T = any> {
  success: boolean;
  message: string;
  dataResponse?: T;
  error?: string;
  stack?: string;
}

export class ResponseUtil {
  static async success<T>(
    res: Response, 
    dataResponse: T, 
    message: string = 'Success', 
    statusCode: number = 200,
    req?: Request
  ): Promise<Response<ApiResponse<T>>> {
    const localizedMessage = req ? await this.getLocalizedMessage(req, message) : message;
    return res.status(statusCode).json({
      success: true,
      message: localizedMessage,
      dataResponse,
    });
  }

  static async error(
    res: Response, 
    message: string = 'Internal Server Error', 
    statusCode: number = 500,
    error?: string,
    stack?: string,
    req?: Request
  ): Promise<Response<ApiResponse>> {
    const localizedMessage = req ? await this.getLocalizedMessage(req, message) : message;
    const response: ApiResponse = {
      success: false,
      message: localizedMessage,
    };

    if (error) response.error = error;
    if (stack && config.NODE_ENV === 'development') {
      response.stack = stack;
    }

    return res.status(statusCode).json(response);
  }

  static async validation(
    res: Response, 
    message: string = 'Validation Error', 
    errors?: any,
    req?: Request
  ): Promise<Response<ApiResponse>> {
    const localizedMessage = req ? await this.getLocalizedMessage(req, message) : message;
    return res.status(400).json({
      success: false,
      message: localizedMessage,
      error: errors,
    });
  }

  private static async getLocalizedMessage(req: Request, message: string): Promise<string> {
    try {
      // Get language from Accept-Language header or query parameter
      const lang = req.query.lang as string || 
                   req.headers['accept-language']?.split(',')[0]?.split('-')[0] || 
                   'en';
      
      // Set lang
      if (lang === 'vi' || lang === 'en') {
        await changeLanguage(lang);
      }
      
      if (message.includes('.')) {
        return t(message);
      }
      
      return message;
    } 
    catch (error) {
      return message;
    }
  }
}

File: servers/DoctorService/Libs/Response.cs
using Microsoft.AspNetCore.Mvc;

namespace DoctorService.Libs
{
    public class ApiResponse<T>
    {
        public bool Success { get; set; }
        public string Message { get; set; } = string.Empty;
        public T? DataResponse { get; set; }
        public string? Error { get; set; }
        public string? Stack { get; set; }
    }

    public class ResponseUtil
    {
        public static ActionResult<ApiResponse<T>> Success<T>(
            T dataResponse,
            string message = "Success",
            int statusCode = 200)
        {
            var response = new ApiResponse<T>
            {
                Success = true,
                Message = message,
                DataResponse = dataResponse
            };

            return new ObjectResult(response)
            {
                StatusCode = statusCode
            };
        }

        public static ActionResult<ApiResponse<T>> Error<T>(
            string message = "Internal Server Error",
            int statusCode = 500,
            string? error = null,
            string? stack = null)
        {
            var response = new ApiResponse<T>
            {
                Success = false,
                Message = message
            };

            if (!string.IsNullOrEmpty(error))
                response.Error = error;

            // Only include stack trace in development environment
            if (!string.IsNullOrEmpty(stack) && 
                Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") == "Development")
            {
                response.Stack = stack;
            }

            return new ObjectResult(response)
            {
                StatusCode = statusCode
            };
        }

        public static ActionResult<ApiResponse<object>> Validation(
            string message = "Validation Error",
            object? errors = null)
        {
            var response = new ApiResponse<object>
            {
                Success = false,
                Message = message,
                Error = errors?.ToString()
            };

            return new ObjectResult(response)
            {
                StatusCode = 400
            };
        }
    }
}

